### DISCUSSION TITLE
Is there something like an API which allows me to write a webapp that can perhaps communicate with maddy to like, reset passwords?

### DISCUSSION THREAD

**Discussion by Damaj301damaj-lol** [emersion/maddy#593] [General]
I have a mail server running maddy(runs great and does the job perfectly), however i have users that want to reset their passwords without telling me their new password directly, is there currently a way to interact with maddy to like change passwords from like an API i can implement in a web page?

**Comment by monban** [comment]
Maybe? The public API of the maddy go module seems like it might have the capability. First, you'd need to load and parse the config file [like the CLI does](https://github.com/foxcpp/maddy/blob/2ddd123c57f0f2fa4eb28ef709ec283dac17da95/internal/cli/ctl/moduleinit.go#L117). Then you can import `github.com/foxcpp/maddy/framework/modules` to instantiate a `PlainUserDB`, then call `SetUserPassword(username, newPassword)` on it. It might be easier to simply shell out to the `maddy` cli, but I do have to wonder... how would you implement something like this? Backup email address? SMS? I think that's the real challenge personally.

**Reply by Damaj301damaj-lol** [reply]
> Maybe? The public API of the maddy go module seems like it might have the capability. First, you'd need to load and parse the config file [like the CLI does](https://github.com/foxcpp/maddy/blob/2ddd123c57f0f2fa4eb28ef709ec283dac17da95/internal/cli/ctl/moduleinit.go#L117). Then you can import `github.com/foxcpp/maddy/framework/modules` to instantiate a `PlainUserDB`, then call `SetUserPassword(username, newPassword)` on it. It might be easier to simply shell out to the `maddy` cli, but I do have to wonder... how would you implement something like this? Backup email address? SMS? I think that's the real challenge personally.

(I solved that first issue using maddy-password-reset)
I was thinking about something like a webgui to manage maddy, such as add emails, remove them, etc

**Comment by foxcpp** [comment]
table.sql_query, table.sql_table modules can be used to read usernames, passwords and so on from any supported SQL database. Such DB in turn can be managed by a simple webapp.

The only issue is keeping it in sync with IMAP accounts. The only currently possible way is to replicate how maddyctl works - "framework" package is considered a public Go API. Config initialization routines are not, but that's the minor part and they are unlikely to change anyway.
